File: kotlin/test/ClientEntityTest.kt
package voxgig.bluefinshieldconexmgmtsdk.sdktest

import java.nio.file.Files
import java.nio.file.Paths

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Assumptions
import org.junit.jupiter.api.Test

import voxgig.bluefinshieldconexmgmtsdk.core.Helpers
import voxgig.bluefinshieldconexmgmtsdk.core.SdkEntity
import voxgig.bluefinshieldconexmgmtsdk.core.BluefinShieldconexMgmtSDK
import voxgig.bluefinshieldconexmgmtsdk.utility.Json
import voxgig.bluefinshieldconexmgmtsdk.utility.struct.Struct

@Suppress("UNCHECKED_CAST", "UNUSED_VARIABLE", "UNUSED_VALUE")
class ClientEntityTest {

  @Test
  fun instance() {
    val testsdk = BluefinShieldconexMgmtSDK.testSDK()
    val ent = testsdk.client(null)
    assertNotNull(ent, "expected non-null client entity")
  }

  @Test
  fun basic() {
    val setup = clientBasicSetup(null)
    // Per-op sdk-test-control.json skip.
    val mode = if (setup.live) "live" else "unit"
    for (op in arrayOf("create", "list", "load", "remove")) {
      val reason = RunnerSupport.skipReason("entityOp", "client.$op", mode)
      Assumptions.assumeTrue(
        reason == null,
        if (reason == null || "" == reason) "skipped via sdk-test-control.json" else reason,
      )
    }
    Assumptions.assumeFalse(
      setup.syntheticOnly,
      "live entity test uses synthetic IDs from fixture — set BLUEFINSHIELDCONEXMGMT_TEST_CLIENT_ENTID JSON to run live",
    )
    val client = setup.client

    // CREATE
    val clientRef01Ent = client.client(null)
    var clientRef01Data: MutableMap<String, Any?> = (Helpers.toMapAny(Struct.getprop(
        Struct.getpath(setup.data, "new.client"), "client_ref01")) ?: linkedMapOf())

    val clientRef01DataResult = clientRef01Ent.create(clientRef01Data, null)
    clientRef01Data = Helpers.toMapAny(clientRef01DataResult) ?: linkedMapOf()
    assertNotNull(clientRef01Data, "expected create result to be a map")
    assertNotNull(clientRef01Data["id"], "expected created entity to have an id")

    // LIST
    val clientRef01Match = linkedMapOf<String, Any?>()

    val clientRef01ListResult = clientRef01Ent.list(clientRef01Match, null)
    assertTrue(clientRef01ListResult is List<*>,
        "expected list result to be an array, got " + clientRef01ListResult)
    val clientRef01List = clientRef01ListResult as List<Any?>

    val foundItem = Struct.select(
        RunnerSupport.entityListToData(clientRef01List),
        Struct.jm("id", clientRef01Data["id"]))
    assertFalse(Struct.isempty(foundItem), "expected to find created entity in list")

    // LOAD
    val clientRef01MatchDt0 = linkedMapOf<String, Any?>()
    clientRef01MatchDt0["id"] = clientRef01Data["id"]
    val clientRef01DataDt0Loaded = clientRef01Ent.load(clientRef01MatchDt0, null)
    val clientRef01DataDt0LoadResult = Helpers.toMapAny(clientRef01DataDt0Loaded) ?: linkedMapOf()
    assertNotNull(clientRef01DataDt0LoadResult, "expected load result to be a map")
    assertEquals(clientRef01Data["id"], clientRef01DataDt0LoadResult["id"],
        "expected load result id to match")

    // REMOVE
    val clientRef01MatchRm0 = linkedMapOf<String, Any?>()
    clientRef01MatchRm0["id"] = clientRef01Data["id"]
    clientRef01Ent.remove(clientRef01MatchRm0, null)

    // LIST
    val clientRef01MatchRt0 = linkedMapOf<String, Any?>()

    val clientRef01ListRt0Result = clientRef01Ent.list(clientRef01MatchRt0, null)
    assertTrue(clientRef01ListRt0Result is List<*>,
        "expected list result to be an array, got " + clientRef01ListRt0Result)
    val clientRef01ListRt0 = clientRef01ListRt0Result as List<Any?>

    val notFoundItem = Struct.select(
        RunnerSupport.entityListToData(clientRef01ListRt0),
        Struct.jm("id", clientRef01Data["id"]))
    assertTrue(Struct.isempty(notFoundItem), "expected removed entity to not be in list")

  }

  @Test
  fun stream() {
    val streamingActive = linkedMapOf<String, Any?>(
      "feature" to linkedMapOf<String, Any?>(
        "streaming" to linkedMapOf<String, Any?>("active" to true),
      ),
    )
    val setup = clientBasicSetup(streamingActive)
    Assumptions.assumeFalse(
      setup.live,
      "stream test streams the seeded fixture data (unit mode only)",
    )

    val ent = setup.client.client(null)
    val match = linkedMapOf<String, Any?>()

    // Materialised list result for the same op.
    val listedResult = ent.list(match, null)
    val listed = (listedResult as? List<Any?>) ?: emptyList<Any?>()

    // stream("list") yields items via the streaming feature's iterator.
    val streamed = ent.stream("list", match, null).toList()
    assertTrue(streamed.size > 0, "expected stream to yield items")
    assertEquals(listed.size, streamed.size, "expected stream to match list count")

    // Fallback: with streaming inactive, stream still yields the materialised
    // items.
    val setup2 = clientBasicSetup(null)
    val ent2 = setup2.client.client(null)
    val streamed2 = ent2.stream("list", match, null).toList()
    assertEquals(listed.size, streamed2.size, "expected fallback stream to match list")
  }

  companion object {
    fun clientBasicSetup(extra: MutableMap<String, Any?>?): RunnerSupport.EntityTestSetup {
      RunnerSupport.loadEnvLocal()

      val entityData: MutableMap<String, Any?>
      try {
        val entityDataSource = Files.readString(Paths.get(
            "..", ".sdk", "test", "entity", "client", "ClientTestData.json"))
        entityData = Helpers.toMapAny(Json.parse(entityDataSource)) ?: linkedMapOf()
      } catch (e: Exception) {
        throw AssertionError("failed to read client test data: " + e.message, e)
      }

      val options = linkedMapOf<String, Any?>()
      options["entity"] = entityData["existing"]

      var client = BluefinShieldconexMgmtSDK.testSDK(options, extra)

      // Generate idmap via transform, matching TS pattern.
      val idnames = mutableListOf<Any?>()
      idnames.add("client01")
      idnames.add("client02")
      idnames.add("client03")
      val idmap = Struct.transform(idnames, Json.parse(
          "{\"`\$PACK`\": [\"\", {" +
          "\"`\$KEY`\": \"`\$COPY`\"," +
          "\"`\$VAL`\": [\"`\$FORMAT`\", \"upper\", \"`\$COPY`\"]" +
          "}]}"))

      // Detect ENTID env override before envOverride consumes it.
      val entidEnvRaw = RunnerSupport.getenv("BLUEFINSHIELDCONEXMGMT_TEST_CLIENT_ENTID")
      val idmapOverridden = entidEnvRaw != null && entidEnvRaw.trim().startsWith("{")

      val envm = linkedMapOf<String, Any?>()
      envm["BLUEFINSHIELDCONEXMGMT_TEST_CLIENT_ENTID"] = idmap
      envm["BLUEFINSHIELDCONEXMGMT_TEST_LIVE"] = "FALSE"
      envm["BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN"] = "FALSE"
      envm["BLUEFINSHIELDCONEXMGMT_APIKEY"] = "NONE"
      val env = RunnerSupport.envOverride(envm)

      var idmapResolved = Helpers.toMapAny(env["BLUEFINSHIELDCONEXMGMT_TEST_CLIENT_ENTID"])
      if (idmapResolved == null) {
        idmapResolved = Helpers.toMapAny(idmap) ?: linkedMapOf()
      }

      val live = "TRUE" == env["BLUEFINSHIELDCONEXMGMT_TEST_LIVE"]
      if (live) {
        val liveOpts = linkedMapOf<String, Any?>()
        liveOpts["apikey"] = env["BLUEFINSHIELDCONEXMGMT_APIKEY"]
        val mergedOpts = Struct.merge(Struct.jt(liveOpts, extra))
        client = BluefinShieldconexMgmtSDK(Helpers.toMapAny(mergedOpts))
      }

      val setup = RunnerSupport.EntityTestSetup()
      setup.client = client
      setup.data = entityData
      setup.idmap = idmapResolved
      setup.env = env
      setup.explain = "TRUE" == env["BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN"]
      setup.live = live
      setup.syntheticOnly = live && !idmapOverridden
      setup.now = System.currentTimeMillis()
      return setup
    }
  }
}

File: rb/test/client_entity_test.rb
# Client entity test

require "minitest/autorun"
require "json"
require_relative "../BluefinShieldconexMgmt_sdk"
require_relative "runner"

class ClientEntityTest < Minitest::Test
  def test_create_instance
    testsdk = BluefinShieldconexMgmtSDK.test(nil, nil)
    ent = testsdk.Client(nil)
    assert !ent.nil?
  end

  # Feature #4: the entity stream(action, ...) method runs the op pipeline and
  # returns an Enumerator over result items. With the streaming feature active
  # it yields the feature's incremental output; otherwise it falls back to the
  # materialised list so stream always yields.
  def test_stream
    seed = {
      "entity" => {
        "client" => {
          "s1" => { "id" => "s1" },
          "s2" => { "id" => "s2" },
          "s3" => { "id" => "s3" },
        },
      },
    }

    # Fallback: streaming inactive -> yields the materialised list items.
    base = BluefinShieldconexMgmtSDK.test(seed, nil)
    seen = base.Client(nil).stream("list", nil, nil).to_a
    assert_equal 3, seen.length

    # Inbound: streaming active -> yields each item from the feature.
    cfg = BluefinShieldconexMgmtConfig.make_config
    if cfg["feature"].is_a?(Hash) && cfg["feature"].key?("streaming")
      sdk = BluefinShieldconexMgmtSDK.test(seed, { "feature" => { "streaming" => { "active" => true } } })
      got = []
      sdk.Client(nil).stream("list", nil, nil).each do |item|
        if item.is_a?(Array)
          got.concat(item)
        else
          got << item
        end
      end
      assert_equal 3, got.length
    end
  end

  def test_basic_flow
    setup = client_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "list", "load", "remove"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "client." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set BLUEFINSHIELDCONEXMGMT_TEST_CLIENT_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    client_ref01_ent = client.Client(nil)
    client_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.client"), "client_ref01"))

    client_ref01_data_result = client_ref01_ent.create(client_ref01_data, nil)
    client_ref01_data = Helpers.to_map(client_ref01_data_result)
    assert !client_ref01_data.nil?
    assert !client_ref01_data["id"].nil?

    # LIST
    client_ref01_match = {}

    client_ref01_list_result = client_ref01_ent.list(client_ref01_match, nil)
    assert client_ref01_list_result.is_a?(Array)

    found_item = Vs.select(
      Runner.entity_list_to_data(client_ref01_list_result),
      { "id" => client_ref01_data["id"] })
    assert !Vs.isempty(found_item)

    # LOAD
    client_ref01_match_dt0 = {
      "id" => client_ref01_data["id"],
    }
    client_ref01_data_dt0_loaded = client_ref01_ent.load(client_ref01_match_dt0, nil)
    client_ref01_data_dt0_load_result = Helpers.to_map(client_ref01_data_dt0_loaded)
    assert !client_ref01_data_dt0_load_result.nil?
    assert_equal client_ref01_data_dt0_load_result["id"], client_ref01_data["id"]

    # REMOVE
    client_ref01_match_rm0 = {
      "id" => client_ref01_data["id"],
    }
    client_ref01_ent.remove(client_ref01_match_rm0, nil)

    # LIST
    client_ref01_match_rt0 = {}

    client_ref01_list_rt0_result = client_ref01_ent.list(client_ref01_match_rt0, nil)
    assert client_ref01_list_rt0_result.is_a?(Array)

    not_found_item = Vs.select(
      Runner.entity_list_to_data(client_ref01_list_rt0_result),
      { "id" => client_ref01_data["id"] })
    assert Vs.isempty(not_found_item)

  end
end

def client_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "client", "ClientTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = BluefinShieldconexMgmtSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["client01", "client02", "client03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["BLUEFINSHIELDCONEXMGMT_TEST_CLIENT_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "BLUEFINSHIELDCONEXMGMT_TEST_CLIENT_ENTID" => idmap,
    "BLUEFINSHIELDCONEXMGMT_TEST_LIVE" => "FALSE",
    "BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN" => "FALSE",
    "BLUEFINSHIELDCONEXMGMT_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["BLUEFINSHIELDCONEXMGMT_TEST_CLIENT_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["BLUEFINSHIELDCONEXMGMT_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["BLUEFINSHIELDCONEXMGMT_APIKEY"],
      },
      extra || {},
    ])
    client = BluefinShieldconexMgmtSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["BLUEFINSHIELDCONEXMGMT_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
